xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +929,691 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
+		t.Errorf("Fraction of RPCs to backend %s: got %v, want %v (margin: +-%v)", backends[2], got, want, errorTolerance)
+	}
+}
+
+// Tests that ring hash policy evaluation will continue past the terminal policy
+// if no results are produced yet.
+func (s) TestRingHash_ContinuesPastTerminalPolicyThatDoesNotProduceResult(t *testing.T) {
+	backends := startTestServiceBackends(t, 2)
+
+	const clusterName = "cluster"
+	endpoints := endpointResource(t, clusterName, backends)
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+
+	route := e2e.DefaultRouteConfig("new_route", "test.server", clusterName)
+
+	// Even though this policy is terminal, since it produces no result, we
+	// continue past it to find a policy that produces results.
+	hashPolicy := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "header_not_present",
+			},
+		},
+	}
+	hashPolicy.Terminal = true
+	hashPolicy2 := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "address_hash",
+			},
+		},
+	}
+	action := route.VirtualHosts[0].Routes[0].Action.(*v3routepb.Route_Route)
+	action.Route.HashPolicy = []*v3routepb.RouteAction_HashPolicy{&hashPolicy, &hashPolicy2}
+
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Note that the first policy does not match, and in the second hash policy,
+	// each type of RPC contains a header value that always hashes to a specific
+	// backend, as the header value matches the value used to create the entry
+	// in the ring.
+	wantBackend := backends[0]
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("address_hash", wantBackend+"_0"))
+	const numRPCs = 100
+	gotPerBackend := checkRPCSendOK(ctx, t, client, numRPCs)
+	if got := gotPerBackend[wantBackend]; got != numRPCs {
+		t.Errorf("Got %v RPCs routed to backend %v, want %v", got, wantBackend, numRPCs)
+	}
+}
+
+// Tests that a random hash is used when header hashing specified a header field
+// that the RPC did not have.
+func (s) TestRingHash_HashOnHeaderThatIsNotPresent(t *testing.T) {
```

Comment:
Same with this test too. What aspect of the ring_hash policy does it test? It seems more to be testing the component that generates the hash value.

Reply:
I expanded the comment a bit.

But OK, clearly the intent of the tests in c-core is to test not just the ring hash load balancing policy, but also its interaction with the component that sets the hash. I think there's value in this, as those interactions are a bit tricky and those end user behaviors are what users ultimately care about.

Do you think we should move those tests somewhere else in the tree?